src/StreamJsonRpc/WebSocketMessageHandler.cs

```diff
@@ -0,0 +1,137 @@
+    using Microsoft;
+
+    /// <summary>
+    /// A message handler for the <see cref="JsonRpc"/> class
+    /// that uses <see cref="System.Net.WebSockets.WebSocket"/> as the transport.
+    /// </summary>
+    public class WebSocketMessageHandler : DelimitedMessageHandler
+    {
+        private readonly ArraySegment<byte> readBuffer;
+        private readonly byte[] writeBuffer;
+        private Decoder readDecoder;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="WebSocketMessageHandler"/> class.
+        /// </summary>
+        /// <param name="webSocket">The <see cref="System.Net.WebSockets.WebSocket"/> used to communicate.</param>
+        /// <param name="bufferSize">
+        /// The size of the buffer to use for reading JSON-RPC messages.
+        /// Messages which exceed this size will be handled properly but may require multiple I/O operations.
+        /// </param>
+        public WebSocketMessageHandler(WebSocket webSocket, int bufferSize = 4096)
+            : base(Stream.Null, Stream.Null, Encoding.UTF8)
+        {
+            Requires.NotNull(webSocket, nameof(webSocket));
+            Requires.Range(bufferSize > 0, nameof(bufferSize));
+
+            this.WebSocket = webSocket;
+            this.readBuffer = new ArraySegment<byte>(new byte[bufferSize]);
+            this.writeBuffer = new byte[bufferSize];
+        }
+
+        /// <summary>
+        /// Gets the <see cref="System.Net.WebSockets.WebSocket"/> used to communicate.
+        /// </summary>
+        public WebSocket WebSocket { get; }
+
+        /// <inheritdoc />
+        protected async override Task<string> ReadCoreAsync(CancellationToken cancellationToken)
+        {
+            WebSocketReceiveResult result = await this.WebSocket.ReceiveAsync(this.readBuffer, cancellationToken).ConfigureAwait(false);
+            if (result.EndOfMessage)
+            {
+                // fast path: the entire message fit within the buffer.
+                return this.Encoding.GetString(this.readBuffer.Array, 0, result.Count);
+            }
+            else
+            {
+                // The message exceeds the size of our buffer.
+                if (this.readDecoder == null)
+                {
+                    this.readDecoder = this.Encoding.GetDecoder();
+                }
+
+                var jsonBuilder = new StringBuilder();
+                char[] decodedChars = new char[this.readBuffer.Array.Length];
+                void DecodeInput()
+                {
+                    int decodedCharsCount = this.readDecoder.GetChars(this.readBuffer.Array, 0, result.Count, decodedChars, 0, result.EndOfMessage);
+                    jsonBuilder.Append(decodedChars, 0, decodedCharsCount);
+                }
+
+                DecodeInput();
+                while (!result.EndOfMessage)
+                {
+                    result = await this.WebSocket.ReceiveAsync(this.readBuffer, cancellationToken).ConfigureAwait(false);
```

Comment:
>(this.readBuffer [](start = 62, length = 16)

are you guarenteed that 2 people won't call ReceiveAsync() on different threads?  Can you share the buffer like this?